### PR TITLE
refactor: feature configs 

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Each feature config has a status field and an optional config which is custom for each feature. This works fine when fetching all feature configs from the feature config endpoint but it gets complicated when receiving an update event when a feature config is updated.

The way `FeatureConfigUpdatedDTO` is defined it's not possible to parse the different feature configs for each feature config type.  Ideally we want to re-use the FeatureConfigDTO defined in `FeatureConfigResponse.kt`.

### Causes

The root problem comes from how the feature config update JSON is structured. 

```
{
  "type": "feature-config.update",
  "name": "mls",
  "data": {
    "status": "enabled",
    "config": { 
      "protocolToggleUsers": ["60368759-d23f-4502-ba6f-68b10e926f7a", ...],
      "defaultProtocol": "proteus",
      "allowedCipherSuites": [1],
      "defaultCipherSuite": 1
    }
  }
}
```

The `name` field is outside the `data` object so it's not possible to know how to parse it without knowing in which context the data object lives, which is unsupported by our JSON parsing library.

### Solutions

Fix the JSON by moving the `name` field into the data object and then use polymorphic JSON parsing feature to parse the different feature configs.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/featureConfigs/FeatureConfigResponse.kt [hunk 1]
```diff
@@ -1,91 +1,185 @@
...
+    @SerialName("selfDeletingMessages")
+    @Serializable
+    data class SelfDeletingMessages(
+        @SerialName("config")
+        val config: SelfDeletingMessagesConfigDTO,
+        @SerialName("status")
+        val status: FeatureFlagStatusDTO
+    ) : FeatureConfigData()
+
+    @SerialName("sndFactorPasswordChallenge")
```
Comment (vitorhugods) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/featureConfigs/FeatureConfigResponse.kt:156]: Is `snd` a typo?

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/notification/EventContentDTO.kt [hunk 2]
```diff
@@ -6,21 +6,81 @@ import com.wire.kalium.network.api.conversation.ConversationMembers
...
 
 @Serializable
 data class EventResponse(
+    @Serializable
     @SerialName("id") val id: String,
     @SerialName("payload") val payload: List<EventContentDTO>?,
     @SerialName("transient") val transient: Boolean = false
 )
 
+object FeatureConfigUpdatedDTOSerializer : KSerializer<EventContentDTO.FeatureConfig.FeatureConfigUpdatedDTO> {
```
Comment (vitorhugods) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/notification/EventContentDTO.kt:39]: Could you please add a bit of documentation here just to describe the purpose of this Serializer + the Json transformer below.

The `type`/`name` thing is fresh in our heads now, but someone else might jump into this code in the future and have no idea why this mechanism is here.

**Review by vitorhugods — APPROVED** [review]
Nicely done 🧑‍🍳🤌🏻

Just a couple of tiny comments.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/featureConfigs/FeatureConfigResponse.kt (hunk 1 above)
Comment (typfel) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/featureConfigs/FeatureConfigResponse.kt:156]: No it's a weird abbreviation used by the backend.

**Pull request merged**